AbstractProcessStage timeouts and cleaner shutdown

laserval opened the pull request:
This is a continuation of @ssimon's work in #250, which fixes #247, #248 and #249.

Processing stages now create a separate processing thread, which can have an optional timeout set. If the processing takes longer than the timeout, the processing thread will be cancelled. If the thread can't cancel, the stage will destroy the JVM to prevent memory leaks and avoid corrupt state. Core should handle this as any other stage crash and restart the stage.

The motivation for this is that there is no way of knowing if a thread will exit if it can't be cancelled, which means it will take up memory and possibly continue doing work. To protect the rest of the stage JVM, the only way to stop the thread is to destroy the whole JVM.

The tests for `AbstractProcessStage` are now improved and no longer rely on timing to pass.

I couldn't come up with a good test for one case, so I created #263 for that.

simonstenstrom committed:
Adds shutdown hook and uncaught exception handler.

Added shutdown hook to Main. Running the shutdown method will terminate all thread started by Hydra. If a thread is hung, Hydra will be killed after 30 seconds

Added an uncaught exception handler that runs the shutdown method if an uncaught exception is thrown by Hydra

If an IOException is caught when trying to start hydra, the shutdown method is called instead of only shutting down the server.

Commit:
Adds possibility to throw a ProcessingException after a configurable timeout, unless the document was processed. If processing times out and cannot be cancelled, the stage will trigger JVM shutdown.

Restructures tests for AbstractProcssStage.

simonstenstrom committed:
Added a stage that gets stuck in an infinite loop and never returns. For example used for testing how your pipeline handles failing libraries that never returns

Added a stage that throws an uncaught runtime exception. Used to test how your pipeline handles unexpected runtime errors

simonstenstrom committed:
Made Main implement a TerminationHandler. Removed all static modifiers in Main. Passed a refrence to main to NodeMaster that passes it to StageRunner

Updated tests to take a mock of TerminationHandler

Commit:
JavaDoc and format cleanup

Cleans up structure and spelling

Removes unused imports

Removed template comments from ConcatenatingHashStage

Commit:
Adds tests for execution timeouts.
Adds StageKiller to allow for tests that check if stages are killed.
Adds TODO about writing tests that check Stage kills.

Commit:
Makes tests for AbstractProcessStage follow the same naming convention

remen commented:
Tests are failing in Travis. Any idea why? Don't want to look to carefully on this until it works. Also, the number of commits here is a bit daunting to code review, especially since a few of them are rewrites of previous ones. Would it make sense to rebase/squash?

laserval commented:
Looking into the test failures.

Good idea about squashing, I'll see what can be done.

laserval commented:
So, apparently the tests failed....randomly? I'll trigger a few more builds and see if there's any pattern to this.

laserval commented:
I've now rebased, squashed and reordered some of the commits to make a cleaner history. I replaced the ProcessThread commit with my Future-implementation, but preserved most of the other changes in separate commits. Hopefully it's easier to follow the changes now.

Commit:
Renames TerminationHandler for consistent naming. Adds note to InfiniteLoopStage to clarify that it can't be interrupted.
